File: app/src/main/java/com/sunnyxibei/gogank/activity/BrowserActivity.java
package com.sunnyxibei.gogank.activity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.sunnyxibei.gogank.R;
import com.sunnyxibei.gogank.global.GlobalConstant;

/**
 * Created by jiayuanbin on 2016/8/7.
 */
public class BrowserActivity extends AppCompatActivity {

    private String title;
    private WebView mWebView;

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_browser);

        Intent intent = getIntent();
        String url = intent.getStringExtra(GlobalConstant.DESC_URL);
        title = intent.getStringExtra(GlobalConstant.TITLE);
        initActionBar();

        mWebView = findViewById(R.id.webview);
        WebSettings webSettings = mWebView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        mWebView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                view.loadUrl(url);
                return true;
            }
        });
        mWebView.loadUrl(url);
    }

    private void initActionBar() {
        ActionBar actionBar = getSupportActionBar();
        assert actionBar != null;
        actionBar.setTitle(title);
        actionBar.setDisplayHomeAsUpEnabled(true);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if ((keyCode == KeyEvent.KEYCODE_BACK) && mWebView.canGoBack()) {
            mWebView.goBack();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: app/src/main/java/com/timeriver/gogank/ui/activity/BrowserActivity.kt
package com.timeriver.gogank.ui.activity

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import com.timeriver.gogank.R

class BrowserActivity : AppCompatActivity() {

    companion object {
        private const val MAX = 100
        private const val BROWSER_URL = "browser_url"

        /**
         * 默认requestCode=-1不提供返回结果
         */
        fun start(activity: Activity, url: String, requestCode: Int = -1) {
            val intent = Intent(activity, BrowserActivity::class.java)
            intent.putExtra(BROWSER_URL, url)
            activity.startActivityForResult(intent, requestCode)
        }
    }

    private var mBrowserUrl: String? = null
    private lateinit var mWebView: WebView

    private lateinit var mProgressBar: ProgressBar
    private lateinit var mWebContainer: FrameLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_browser)
        mProgressBar = findViewById(R.id.progress_bar)
        mWebContainer = findViewById(R.id.fl_web_container)

        initData()
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initData() {
        mBrowserUrl = intent.getStringExtra(BROWSER_URL)

        mWebView = WebView(this)
        mWebContainer.addView(mWebView)
        mWebView.requestFocus()
        mWebView.isHorizontalScrollBarEnabled = false
        mWebView.isVerticalScrollBarEnabled = false
        mWebView.overScrollMode = WebView.OVER_SCROLL_NEVER
        val webSettings = mWebView.settings
        webSettings.javaScriptEnabled = true
        webSettings.useWideViewPort = true
        webSettings.loadWithOverviewMode = true
        webSettings.domStorageEnabled = true
        webSettings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        mWebView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                if (url.startsWith("http:") || url.startsWith("https:")) {
                    return false
                }
                // Otherwise allow the OS to handle things like tel, mailto, etc.
                try {
                    val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                    startActivity(intent)
                } catch (e: ActivityNotFoundException) {
                    e.printStackTrace()
                }
                return true
            }
        }
        mWebView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView, newProgress: Int) {
                mProgressBar.progress = newProgress
                if (newProgress == MAX) {
                    mProgressBar.visibility = View.GONE
                } else {
                    mProgressBar.visibility = View.VISIBLE
                }
            }
        }
        mWebView.loadUrl(mBrowserUrl)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && mWebView.canGoBack()) {
            mWebView.goBack()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        mWebView.removeAllViews()
        mWebView.clearHistory()
        mWebView.clearCache(true)
        mWebView.destroy()
    }
}